### PR TITLE
[Brewmaster] Unit tests

### PR DESCRIPTION
**Pull request by shighman**
Funny comment...
Minor change to 4pc to use constants over name
Added unit tests for the BrM modules which do work. All pass but for future refactoring.

I didn't change the changelog because there is no difference in the code, if you would like me to then i can.

### CODE REVIEW

File: src/tests/Parser/Brewmaster/SimpleFIght.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+const thisPlayer = 1;
+const enemy = 2;
+
+export function processEvents(events, module, fightEndTime) {
+  if (!fightEndTime) {
+    fightEndTime = FIGHT_END;
+  }
+  events.forEach(event => {
+    if (event.timestamp <= fightEndTime) {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/SimpleFIght.js]: You should use fireEvent instead; it's also being tested iirc so it should always work properly and make for cleaner test code.

File: src/tests/Parser/Brewmaster/SimpleFight.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+const thisPlayer = 1;
+const enemy = 2;
+
+export function processEvents(events, module, fightEndTime) {
+  if (!fightEndTime) {
+    fightEndTime = FIGHT_END;
+  }
+  events.forEach(event => {
+    if (event.timestamp <= fightEndTime) {
+      if (event.type === "absorbed") {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/SimpleFight.js]: Please use the `fireEvent` function of the module for firing the events. iirc it's also being tested so it should work reliably enough. This reduces the amount of setup code needed in tests.

**Comment by shighman** [comment]
Sorry so used to not being able to do Setup and Teardown methods, i just don't use them any more out of habit.

**Comment by MartijnHols** [comment]
Have you seen the comments on Discord?

**Comment by shighman** [comment]
I have made some changes, first i changed to beforeAll but changed to beforeEach, they are missing some edge cases i need to add them i will do some more after work today.

File: src/tests/Parser/Brewmaster/DamageTaken.test.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import DamageTaken from 'Parser/BrewmasterMonk/Modules/Core/DamageTaken';
+import { events, processEvents } from './Fixtures/SimpleFight';
+
+const damageTaken = new DamageTaken();
+
+describe('Brewmaster.DamageTaken', () => {
+  beforeAll(() => {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/DamageTaken.test.js]: This should be beforeEach, resetting state before each test is best practice

File: src/tests/Parser/Brewmaster/Stagger.test.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import Stagger from 'Parser/BrewmasterMonk/Modules/Core/Stagger';
+import { events, processEvents, FIGHT_END, TOTAL_STAGGERED } from './Fixtures/SimpleFight';
+
+const stagger = new Stagger();
+
+describe('Brewmaster.Stagger', () => {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/Stagger.test.js]: I think writing the test cases as if they're requirements would make for better tests, I don't think you should try to only have 1 assertion per test case, there's no gain in doing so. I'd propose the follow test cases:

- it *tracks the total amount of damage staggered*
- it *tracks the total amount of stagger damage taken*
- it *tracks the amount of stagger missing from the fight*

File: src/tests/Parser/Brewmaster/Stagger.test.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    staggerLocal.owner = myOwner;
+    expect(staggerLocal.owner.fight.end_time).toBe(FIGHT_END);
+  });
+  it('How much damage in total is absorbed by stagger', () => {
+    expect(stagger.totalPhysicalStaggered + stagger.totalMagicalStaggered).toBe(599);
+  });
+  it('How much physical damage was staggered', () => {
+    expect(stagger.totalPhysicalStaggered).toBe(300);
+  });
+  it('When was the last non stagger damage event', () => {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/Stagger.test.js]: To what purpose is this being tracked?

File: src/tests/Parser/Brewmaster/T20_2pc.test.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import T20_2pc from 'Parser/BrewmasterMonk/Modules/Items/T20_2pc';
+import { events, processEvents } from './Fixtures/SimpleFight';
+
+const item = new T20_2pc();
+
+describe('Brewmaster.T20_2pc', () => {
+  beforeAll(() => {
+    processEvents(events, item);
+  });
+  it('How many orbs spawned', () => {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/T20_2pc.test.js]: - it *tracks the amount of orb spawns*

File: src/tests/Parser/Brewmaster/T20_2pc.test.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    expect(item.orbTriggeredBy2Pc).toBe(1);
+  });
+  it('How many brews were used which could spawn an orb', () => {
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/T20_2pc.test.js]: For what purpose is this being tracked here? Shouldn't this be private?

File: src/Parser/BrewmasterMonk/Modules/Spells/IronSkinBrew.js [hunk 8]
```diff
@@ -17,9 +17,6 @@ class IronSkinBrew extends Module {
   hitsWithoutIronSkinBrew = 0;
   damageWithoutIronSkinBrew = 0;
 
-  // This value is just for dev purposes to check the total, its not shown on the tooltip.
```
Comment (MartijnHols) [inline src/Parser/BrewmasterMonk/Modules/Spells/IronSkinBrew.js]: If it's for dev purposes you can add a dev statement, but please don't test it, it makes it feels like it's public API that is required, making it hard for someone else / future you from coming back and refactoring this as they won't understand the purpose of this easily

File: src/tests/Parser/Brewmaster/DamageTaken.test.js [hunk 9]
```diff
@@ -1,10 +1,11 @@
 import DamageTaken from 'Parser/BrewmasterMonk/Modules/Core/DamageTaken';
 import { events, processEvents } from './Fixtures/SimpleFight';
 
-const damageTaken = new DamageTaken();
+let damageTaken = null;
```
Comment (MartijnHols) [inline src/tests/Parser/Brewmaster/DamageTaken.test.js]: This could be put in the `describe` function, and can be just `let damageTaken;` without value

File: src/tests/Parser/Brewmaster/Fixtures/SimpleFight.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import SPELLS from 'common/SPELLS';
+
+export const FIGHT_END = 11000;
+export const TOTAL_STAGGERED = 240;
+
+const thisPlayer = 1;
+const enemy = 2;
+
+export function processEvents(events, module, fightEndTime) {
```
Comment (shighman) [inline src/tests/Parser/Brewmaster/Fixtures/SimpleFight.js]: I have moved it out into its own file because i need to have more than one fight for edge cases, i'll maybe look at moving it later though.

File: src/tests/Parser/Brewmaster/Fixtures/SimpleFight.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  events.forEach(event => {
+    if (event.timestamp <= fightEndTime) {
+      module.triggerEvent('toPlayer_' + event.type, event);
+      module.triggerEvent('byPlayer_' + event.type, event);
+    }
+  });
+}
+
+export const events = [
+  { type: "damage", sourceid: enemy, targetid: thisPlayer, amount: 400, timestamp: 1, ability: { guid: 1 } },
```
Comment (shighman) [inline src/tests/Parser/Brewmaster/Fixtures/SimpleFight.js]: I have renamed that one as SimpleFight as the name implies, and will create another array in another file to test other scenarios.

File: src/Parser/BrewmasterMonk/Modules/Spells/IronSkinBrew.js (hunk 8 above)
Comment (shighman) [inline src/Parser/BrewmasterMonk/Modules/Spells/IronSkinBrew.js]: It doesn't actually need to be there now since i have automated tests, i shouldn't have included it in the tests.

File: src/tests/Parser/Brewmaster/T20_2pc.test.js (hunk 7 above)
Comment (shighman) [inline src/tests/Parser/Brewmaster/T20_2pc.test.js]: This is used to calculate the actual proc rate of the T20 2pc, each brew has a 40% chance

**Comment by MartijnHols** [comment]
Thanks!
